crates/ide_assists/src/handlers/unnecessary_async.rs

```diff
@@ -0,0 +1,270 @@
+        check_assist(
+            unnecessary_async,
+            concat!(
+                "pub async fn f4() { }\n",
+                "mod a { pub async f$0n f2() { } }\n",
+                "pub async fn f() { a::f2().await }\n",
+                "pub async fn f3() { a::f2().await }"
+            ),
+            concat!(
+                "pub async fn f4() { }\n",
+                "mod a { pub  fn f2() { } }\n",
+                "pub async fn f() { a::f2() }\n",
+                "pub async fn f3() { a::f2() }"
+            ),
+        )
+    }
+
+    #[test]
+    fn applies_and_removes_await_on_inner_await() {
+        check_assist(
+            unnecessary_async,
+            concat!(
+                "pub async fn f() { f2().await }\n",
+                "pub async f$0n f2() -> i32 { 1 }\n",
+                // Ensure that it is the first await below that is removed
+                "pub async fn f3() { f4(f2().await).await }",
+                "pub async fn f4(i: i32) { }\n",
+            ),
+            concat!(
+                "pub async fn f() { f2() }\n",
+                "pub  fn f2() -> i32 { 1 }\n",
+                "pub async fn f3() { f4(f2()).await }",
+                "pub async fn f4(i: i32) { }\n",
+            ),
+        )
+    }
+
+    #[test]
+    fn applies_and_removes_await_on_outer_await() {
+        check_assist(
+            unnecessary_async,
+            concat!(
+                "pub async fn f() { f2().await }\n",
+                "pub async f$0n f2(i: i32) { }\n",
+                // Ensure that it is the second await below that is removed
+                "pub async fn f3() { f2(f4().await).await }",
+                "pub async fn f4() -> i32 { 1 }\n",
+            ),
+            concat!(
+                "pub async fn f() { f2() }\n",
+                "pub  fn f2(i: i32) { }\n",
+                "pub async fn f3() { f2(f4().await) }",
+                "pub async fn f4() -> i32 { 1 }\n",
+            ),
+        )
+    }
+
+    #[test]
+    fn applies_on_method_call() {
+        check_assist(
+            unnecessary_async,
+            concat!(
+                "pub struct S { }\n",
+                "impl S { pub async f$0n f2(&self) { } }\n",
+                "pub async fn f(s: &S) { s.f2().await }\n",
+            ),
+            concat!(
+                "pub struct S { }\n",
+                "impl S { pub  fn f2(&self) { } }\n",
+                "pub async fn f(s: &S) { s.f2() }\n",
+            ),
+        )
+    }
+
+    #[test]
+    fn applies_through_macros() {
+        check_assist(
+            unnecessary_async,
+            concat!(
+                "macro_rules! m { () => { f2().await }; }\n",
+                "pub async f$0n f2() { }\n",
+                "pub async fn f() { m!() }\n",
+            ),
+            concat!(
+                "macro_rules! m { () => { f2() }; }\n",
+                "pub  fn f2() { }\n",
+                "pub async fn f() { m!() }\n",
+            ),
+        )
+    }
```

Comment:
I would have expected this test to pass. I may be wrong though. If it is not meant to pass, I just have to remove it. All other tests pass.

Reply:
Ye you can remove this. Rewriting macro definitions is a problem that we'll most likely never be able to properly solve, see https://rust-analyzer.github.io/blog/2020/03/30/macros-vs-rename.html for some more information on that problem :)

Reply:
That's an interesting read!